Unable to build Python on macOS Big Sur with Xcode 12 beta

I installed macOS Big Sur beta 1, Xcode 12.0 beta (12A6159) on an old Mac to see what breaks. I'm able to install PyEnv via Homebrew, but I'm not able to build any Python version.

Apple clang version 12.0.0

```
% pyenv install 3.8.3
python-build: use openssl@1.1 from homebrew
python-build: use readline from homebrew
Downloading Python-3.8.3.tar.xz...
-> https://www.python.org/ftp/python/3.8.3/Python-3.8.3.tar.xz
Installing Python-3.8.3...
python-build: use readline from homebrew
python-build: use zlib from xcode sdk

BUILD FAILED (OS X 10.16 using python-build 20180424)

Inspect or clean up the working tree at /var/folders/sl/5j1zvmtj0gb8qpgkklkf38dh0000gr/T/python-build.20200626170507.14625
Results logged to /var/folders/sl/5j1zvmtj0gb8qpgkklkf38dh0000gr/T/python-build.20200626170507.14625.log

Last 10 log lines:
./Modules/posixmodule.c:9141:12: note: forward declaration of 'struct sf_hdtr'
    struct sf_hdtr sf;
           ^
./Modules/posixmodule.c:9221:15: error: implicit declaration of function 'sendfile' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
        ret = sendfile(in, out, offset, &sbytes, &sf, flags);
              ^
2 errors generated.
make: *** [Modules/posixmodule.o] Error 1
make: *** Waiting for unfinished jobs....
1 warning generated.
```

[pyenv_install.log](https://github.com/pyenv/pyenv/files/4837680/pyenv_install.log)

Note that `MacOSX.sdk` is symlinked to `MacOSX10.16.sdk`. Pointing the symlink to MacOSX10.15.sdk doesn't help.

I didn't try switching to Xcode 11 yet, though that will probably help, if it's indeed a clang problem.

Downloading and installing Python 3.8.3 from the website works fine (python3 executable).

I am also facing similar issue with `posixmodule`
I have tried following options unsuccessfully :(

1. Setting SDKROOT and MACOSX_DEPLOYMENT_TARGET
> SDKROOT=/Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk \
MACOSX_DEPLOYMENT_TARGET=10.16 \
PYTHON_CONFIGURE_OPTS="--enable-framework" \
pyenv install  3.8.0

2. Setting CFLAGS AND LDFLAGS
> CFLAGS="-I$(brew --prefix openssl)/include -I$(brew --prefix readline)/include -I$(xcrun --show-sdk-path)/usr/include" LDFLAGS="-L$(brew --prefix openssl)/lib -L$(brew --prefix readline)/lib -L$(brew --prefix zlib)/lib" pyenv install 3.8.0

```console
% sw_vers 
ProductName:    macOS
ProductVersion: 10.16
BuildVersion:   20A4300b
```

```console
% clang --version
Apple clang version 12.0.0 (clang-1200.0.22.7)
Target: x86_64-apple-darwin20.0.0
Thread model: posix
InstalledDir: /Applications/Xcode-beta.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

```console
% pyenv install 3.8.0
python-build: use openssl@1.1 from homebrew
python-build: use readline from homebrew
Installing Python-3.8.0...
python-build: use readline from homebrew
python-build: use zlib from xcode sdk

BUILD FAILED (OS X 10.16 using python-build 20180424)

Inspect or clean up the working tree at /var/folders/xw/srzcqq0j1lqfsm1f23wqp6380000gn/T/python-build.20200708180922.77980
Results logged to /var/folders/xw/srzcqq0j1lqfsm1f23wqp6380000gn/T/python-build.20200708180922.77980.log

Last 10 log lines:
           ^
./Modules/posixmodule.c:9084:15: error: implicit declaration of function 'sendfile' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
        ret = sendfile(in, out, offset, &sbytes, &sf, flags);
              ^
clang -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -I/Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include   -I/Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include   -std=c99 -Wextra -Wno-unused-result -Wno-unused-parameter -Wno-missing-field-initializers -Wstrict-prototypes -Werror=implicit-function-declaration  -I./Include/internal  -I. -I./Include -I/usr/local/opt/readline/include -I/usr/local/opt/readline/include -I/Users/pritam/.pyenv/versions/3.8.0/include  -I/usr/local/opt/readline/include -I/usr/local/opt/readline/include -I/Users/pritam/.pyenv/versions/3.8.0/include   -DPy_BUILD_CORE_BUILTIN  -c ./Modules/pwdmodule.c -o Modules/pwdmodule.o
clang -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -I/Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include   -I/Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include   -std=c99 -Wextra -Wno-unused-result -Wno-unused-parameter -Wno-missing-field-initializers -Wstrict-prototypes -Werror=implicit-function-declaration  -I./Include/internal  -I. -I./Include -I/usr/local/opt/readline/include -I/usr/local/opt/readline/include -I/Users/pritam/.pyenv/versions/3.8.0/include  -I/usr/local/opt/readline/include -I/usr/local/opt/readline/include -I/Users/pritam/.pyenv/versions/3.8.0/include   -DPy_BUILD_CORE_BUILTIN  -c ./Modules/_sre.c -o Modules/_sre.o
2 errors generated.
make: *** [Modules/posixmodule.o] Error 1
make: *** Waiting for unfinished jobs....
1 warning generated.
```

[pyenv_3_8_0_install.log](https://github.com/pyenv/pyenv/files/4890681/pyenv_3_8_0_install.log)

I tried building using those inputs to my .zshrc and still got an error:

`andytriboletti@andys-mac ~ % CFLAGS="-I$(brew --prefix openssl)/include -I$(brew --prefix bzip2)/include -I$(brew --prefix readline)/include -I$(xcrun --show-sdk-path)/usr/include" LDFLAGS="-L$(brew --prefix openssl)/lib -L$(brew --prefix readline)/lib -L$(brew --prefix zlib)/lib -L$(brew --prefix bzip2)/lib" \
pyenv install --patch 3.8.3 < <(curl -sSL https://github.com/python/cpython/commit/8ea6353.patch\?full_index\=1)

python-build: use openssl@1.1 from homebrew
python-build: use readline from homebrew
Downloading Python-3.8.3.tar.xz...
-> https://www.python.org/ftp/python/3.8.3/Python-3.8.3.tar.xz
Installing Python-3.8.3...
patching file Misc/NEWS.d/next/macOS/2020-06-24-13-51-57.bpo-41100.mcHdc5.rst
patching file configure
Hunk #1 succeeded at 3398 (offset -28 lines).
patching file configure.ac
Hunk #1 succeeded at 498 (offset -12 lines).
python-build: use readline from homebrew
python-build: use zlib from xcode sdk

BUILD FAILED (OS X 11.0 using python-build 20180424)

Inspect or clean up the working tree at /var/folders/1r/bdbwv9r53ls_80hpk1t6fvlm0000gn/T/python-build.20200715094722.77820
Results logged to /var/folders/1r/bdbwv9r53ls_80hpk1t6fvlm0000gn/T/python-build.20200715094722.77820.log

Last 10 log lines:
checking size of _Bool... 1
checking size of off_t... 8
checking whether to enable large file support... no
checking size of time_t... 8
checking for pthread_t... yes
checking size of pthread_t... 8
checking size of pthread_key_t... 8
checking whether pthread_key_t is compatible with int... no
configure: error: Unexpected output of 'arch' on OSX
make: *** No targets specified and no makefile found.  Stop.
andytriboletti@andys-mac ~ % arch
arm64
`

A unique issue may occur if one has installed and set the latest `XCode` SDK.  `pyenv` fails on run due to invalid clang version:

```bash
  λ sudo xcode-select --switch /Applications/Xcode.app/
  λ clang --version
Apple clang version 11.0.3 (clang-1103.0.32.62)
Target: x86_64-apple-darwin20.1.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
  λ ./pyenv_install_rdbreak_original_script.sh
```

```bash
BUILD FAILED (OS X 11.0 using python-build 20180424)

Inspect or clean up the working tree at /var/folders/4m/1v3nfl_x32z07msw_m6y_1m40000gn/T/python-build.20200919155415.61074
Results logged to /var/folders/4m/1v3nfl_x32z07msw_m6y_1m40000gn/T/python-build.20200919155415.61074.log

Last 10 log lines:
checking for python3.8... python3.8
checking for --enable-universalsdk... no
checking for --with-universal-archs... no
checking MACHDEP... "darwin"
checking for gcc... clang
checking whether the C compiler works... no
configure: error: in `/var/folders/4m/1v3nfl_x32z07msw_m6y_1m40000gn/T/python-build.20200919155415.61074/Python-3.8.5':
configure: error: C compiler cannot create executables
See `config.log' for more details
make: *** No targets specified and no makefile found.  Stop.
Done!
```

The issue relies on the fact that Big Sur is now serving _both_ versions `11.x` and `10.x` by default (for backwards compatibility) with automatic context switch ([source](https://twitter.com/jeremyhu/status/1285230460402987008)):

![rfXPY](https://user-images.githubusercontent.com/2829082/93688029-c379de80-fa90-11ea-852e-b4cfa6f37e96.png)

There are two viable solutions:

#### First Solution

Set your default SDK back to the system's, which will automatically use context switching and utilize SDK `10.x` with `pyenv`:

```bash
  λ sudo xcode-select --switch /Library/Developer/CommandLineTools
```
 
#### Second Solution

The second solution involves simply updating @rdbreak's script above to include the new `SYSTEM_VERSION_COMPAT=1` flag: 

```bash
#!/bin/bash
echo "Installing Dependencies"
brew install zlib
brew install sqlite
brew install bzip2
brew install libiconv
brew install libzip
echo "Done!"
echo
echo -e "Setting Environment Variables"
export SYSTEM_VERSION_COMPAT=1
export LDFLAGS="${LDFLAGS} -L/usr/local/opt/zlib/lib"
export CPPFLAGS="${CPPFLAGS} -I/usr/local/opt/zlib/include"
export LDFLAGS="${LDFLAGS} -L/usr/local/opt/sqlite/lib"
export CPPFLAGS="${CPPFLAGS} -I/usr/local/opt/sqlite/include"
export PKG_CONFIG_PATH="${PKG_CONFIG_PATH} /usr/local/opt/zlib/lib/pkgconfig"
export PKG_CONFIG_PATH="${PKG_CONFIG_PATH} /usr/local/opt/sqlite/lib/pkgconfig"
echo "Done!"
echo
echo "Installing Python 3.8.5"
pyenv install 3.8.5
echo "Done!"
exit 0
```

Either solution will work in the meantime.

Hello everyone, I'm trying to install python 3.6.9 using @synchronizing script however this is the error I ran into
`Last 10 log lines:
clang -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -I/Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include   -I/Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include   -std=c99 -Wextra -Wno-unused-result -Wno-unused-parameter -Wno-missing-field-initializers -Wstrict-prototypes   -I. -I./Include -I/usr/local/opt/readline/include -I/usr/local/opt/readline/include -I/usr/local/opt/openssl@1.1/include -I/Users/henrychea/.pyenv/versions/3.6.9/include  -I/usr/local/opt/zlib/include -I/usr/local/opt/sqlite/include -I/usr/local/opt/readline/include -I/usr/local/opt/readline/include -I/usr/local/opt/openssl@1.1/include -I/Users/henrychea/.pyenv/versions/3.6.9/include  -I/usr/local/opt/zlib/include -I/usr/local/opt/sqlite/include  -DPy_BUILD_CORE  -c ./Modules/errnomodule.c -o Modules/errnomodule.o
./Modules/posixmodule.c:8210:15: error: implicit declaration of function 'sendfile' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
        ret = sendfile(in, out, offset, &sbytes, &sf, flags);
              ^
./Modules/posixmodule.c:10432:5: warning: code will never be executed [-Wunreachable-code]
    Py_FatalError("abort() called from Python code didn't abort!");
    ^~~~~~~~~~~~~
1 warning and 1 error generated.
make: *** [Modules/posixmodule.o] Error 1
make: *** Waiting for unfinished jobs....
Done!
`
Currently I'm on macOS 11.00 (20A5374i) Beta 8. This issue isn't present when installing python 3.8.5

So I have a weird situation pyenv will install some python versions but not all. See below. I am on Big Sur Beta also. For the errors for the failed install I get:
error: implicit declaration of function 'sendfile' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
        ret = sendfile(in, out, offset, &sbytes, &sf, flags);
              ^
1 error generated.
make: *** [Modules/posixmodule.o] Error 1
make: *** Waiting for unfinished jobs....
1 warning generated.
 
For example:
pyenv install 3.6.x - Fails
pyenv install 3.7.x - Fails
pyenv install 3.8.3 - Fails
pyenv install 3.8.4-6 - Success
pyenv install 3.9.0 - Success

can't build 3.6.X too with 

`Last 10 log lines:
./Modules/posixmodule.c:8210:15: error: implicit declaration of function 'sendfile' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
        ret = sendfile(in, out, offset, &sbytes, &sf, flags);
              ^
./Modules/posixmodule.c:10432:5: warning: code will never be executed [-Wunreachable-code]
    Py_FatalError("abort() called from Python code didn't abort!");
    ^~~~~~~~~~~~~
`

upper versions looks fine. 
but 3.6 won't work as installed with installer, wont build from python.org src, won't build with pyenv... =(

There's at least a couple problems in this ticket so I'll try to summarize what worked and didn't for me upgrading from working versions of the same Pythons in Catalina.

The following worked for 2.7.18, 3.7.9, 3.8.6, and 3.9.0:

```console
$ LDFLAGS="-L$(xcrun --show-sdk-path)/usr/lib" pyenv install 3.6.8
```

However, on 3.7.6 there's different errors based on whether you use SYSTEM_VERSION_COMPAT, showing just the default here:

```console
$ LDFLAGS="-L$(xcrun --show-sdk-path)/usr/lib" pyenv install 3.7.6
... snip ...
Last 10 log lines:
./Modules/posixmodule.c:8436:15: error: implicit declaration of function 'sendfile' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
        ret = sendfile(in, out, offset, &sbytes, &sf, flags);
```

3.6.12:

```console
$ LDFLAGS="-L$(xcrun --show-sdk-path)/usr/lib" pyenv install 3.6.12
... snip ...
Last 10 log lines:
        ret = sendfile(in, out, offset, &sbytes, &sf, flags);
              ^
./Modules/posixmodule.c:10432:5: warning: code will never be executed [-Wunreachable-code]
    Py_FatalError("abort() called from Python code didn't abort!");
    ^~~~~~~~~~~~~
```

3.5.10:

```console
$ LDFLAGS="-L$(xcrun --show-sdk-path)/usr/lib" pyenv install 3.5.10
... snip ...
Last 10 log lines:
    ^~~~~~~~~~~~~
./Modules/posixmodule.c:11345:9: warning: code will never be executed [-Wunreachable-code]
        posix_error();
        ^~~~~~~~~~~
```

It does always warn me: `WARNING: The Python readline extension was not compiled. Missing the GNU readline lib?`

Python 3.8.3 still results in the same error as before, but Python 3.8.6 does build so that's good enough for me.

More annoying is that Python 3.6.12 doesn't build:

```
% pyenv install 3.6.12
python-build: use openssl@1.1 from homebrew
python-build: use readline from homebrew
Downloading Python-3.6.12.tar.xz...
-> https://www.python.org/ftp/python/3.6.12/Python-3.6.12.tar.xz
Installing Python-3.6.12...
python-build: use readline from homebrew
python-build: use zlib from xcode sdk

BUILD FAILED (OS X 11.0.1 using python-build 1.2.21-1-g943015eb)

Inspect or clean up the working tree at /var/folders/dq/pp76yn792zn8hpdcsv5t4b040000gn/T/python-build.20201116174329.85347
Results logged to /var/folders/dq/pp76yn792zn8hpdcsv5t4b040000gn/T/python-build.20201116174329.85347.log

Last 10 log lines:
./Modules/posixmodule.c:8210:15: error: implicit declaration of function 'sendfile' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
        ret = sendfile(in, out, offset, &sbytes, &sf, flags);
              ^
clang -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include   -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include   -std=c99 -Wextra -Wno-unused-result -Wno-unused-parameter -Wno-missing-field-initializers -Wstrict-prototypes   -I. -I./Include -I/usr/local/opt/readline/include -I/usr/local/opt/readline/include -I/usr/local/opt/openssl@1.1/include -I/Users/sjors/.pyenv/versions/3.6.12/include  -I/usr/local/opt/readline/include -I/usr/local/opt/readline/include -I/usr/local/opt/openssl@1.1/include -I/Users/sjors/.pyenv/versions/3.6.12/include   -DPy_BUILD_CORE  -c ./Modules/errnomodule.c -o Modules/errnomodule.o
./Modules/posixmodule.c:10432:5: warning: code will never be executed [-Wunreachable-code]
    Py_FatalError("abort() called from Python code didn't abort!");
    ^~~~~~~~~~~~~
1 warning and 1 error generated.
make: *** [Modules/posixmodule.o] Error 1
make: *** Waiting for unfinished jobs....
```

I reinstalled everything (I hope) but it is not working for me.

- macOS 11.0.1
- Developertools for XCode 12.3 beta
- python 2.7.16 (system)
- pip 20.2.4 (istalled in system's python)

sudo xcode-select --switch /Library/Developer/CommandLineTools

Both commands 
pyenv install 3.9.0
SYSTEM_VERSION_COMPAT=1 pyenv install 3.9.0

result in:

```
python-build: use openssl@1.1 from homebrew
python-build: use readline from homebrew
Downloading Python-3.9.0.tar.xz...
-> https://www.python.org/ftp/python/3.9.0/Python-3.9.0.tar.xz
Installing Python-3.9.0...
python-build: use readline from homebrew
python-build: use zlib from xcode sdk

BUILD FAILED (OS X 10.16 using python-build 20180424)

Inspect or clean up the working tree at /var/folders/0b/spth3nrn4cb2zvnmppd5wmdw0000gn/T/python-build.20201116182443.71956
Results logged to /var/folders/0b/spth3nrn4cb2zvnmppd5wmdw0000gn/T/python-build.20201116182443.71956.log

Last 10 log lines:
  File "/private/var/folders/0b/spth3nrn4cb2zvnmppd5wmdw0000gn/T/python-build.20201116182443.71956/Python-3.9.0/Lib/ensurepip/__init__.py", line 210, in _main
    return _bootstrap(
  File "/private/var/folders/0b/spth3nrn4cb2zvnmppd5wmdw0000gn/T/python-build.20201116182443.71956/Python-3.9.0/Lib/ensurepip/__init__.py", line 129, in _bootstrap
    return _run_pip(args + [p[0] for p in _PROJECTS], additional_paths)
  File "/private/var/folders/0b/spth3nrn4cb2zvnmppd5wmdw0000gn/T/python-build.20201116182443.71956/Python-3.9.0/Lib/ensurepip/__init__.py", line 38, in _run_pip
    return subprocess.run([sys.executable, "-c", code], check=True).returncode
  File "/private/var/folders/0b/spth3nrn4cb2zvnmppd5wmdw0000gn/T/python-build.20201116182443.71956/Python-3.9.0/Lib/subprocess.py", line 524, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['/private/var/folders/0b/spth3nrn4cb2zvnmppd5wmdw0000gn/T/python-build.20201116182443.71956/Python-3.9.0/python.exe', '-c', '\nimport runpy\nimport sys\nsys.path = [\'/var/folders/0b/spth3nrn4cb2zvnmppd5wmdw0000gn/T/tmpux9blbc2/setuptools-49.2.1-py3-none-any.whl\', \'/var/folders/0b/spth3nrn4cb2zvnmppd5wmdw0000gn/T/tmpux9blbc2/pip-20.2.3-py2.py3-none-any.whl\'] + sys.path\nsys.argv[1:] = [\'install\', \'--no-cache-dir\', \'--no-index\', \'--find-links\', \'/var/folders/0b/spth3nrn4cb2zvnmppd5wmdw0000gn/T/tmpux9blbc2\', \'--root\', \'/\', \'--upgrade\', \'setuptools\', \'pip\']\nrunpy.run_module("pip", run_name="__main__", alter_sys=True)\n']' returned non-zero exit status 1.
make: *** [install] Error 1
```

Hi Felipe, 

> pyenv install --patch 3.8.0 <<(curl -sSL https://github.com/python/cpython/commit/8ea6353.patch\?full_index\=1)

thanks for this but - in my case - this seams not to be the problem. I can do the build with the patch, but the error is then:

```
CFLAGS="-I$(brew --prefix openssl)/include -I$(brew --prefix bzip2)/include -I$(brew --prefix readline)/include -I$(xcrun --show-sdk-path)/usr/include" LDFLAGS="-L$(brew --prefix openssl)/lib -L$(brew --prefix readline)/lib -L/usr/local/lib -L$(brew --prefix bzip2)/lib" pyenv install --patch 3.8.0 < <(curl -sSL https://github.com/python/cpython/commit/8ea6353.patch\?full_index\=1)
python-build: use openssl@1.1 from homebrew
python-build: use readline from homebrew
Downloading Python-3.8.0.tar.xz...
-> https://www.python.org/ftp/python/3.8.0/Python-3.8.0.tar.xz
Installing Python-3.8.0...
patching file Misc/NEWS.d/next/macOS/2020-06-24-13-51-57.bpo-41100.mcHdc5.rst
patching file configure
Hunk #1 succeeded at 3394 (offset -32 lines).
patching file configure.ac
Hunk #1 succeeded at 498 (offset -12 lines).
python-build: use readline from homebrew
python-build: use zlib from xcode sdk

BUILD FAILED (OS X 11.0.1 using python-build 20180424)

Inspect or clean up the working tree at /var/folders/0b/spth3nrn4cb2zvnmppd5wmdw0000gn/T/python-build.20201116195357.15726
Results logged to /var/folders/0b/spth3nrn4cb2zvnmppd5wmdw0000gn/T/python-build.20201116195357.15726.log

Last 10 log lines:
    return _bootstrap(
  File "/private/var/folders/0b/spth3nrn4cb2zvnmppd5wmdw0000gn/T/python-build.20201116195357.15726/Python-3.8.0/Lib/ensurepip/__init__.py", line 119, in _bootstrap
    return _run_pip(args + [p[0] for p in _PROJECTS], additional_paths)
  File "/private/var/folders/0b/spth3nrn4cb2zvnmppd5wmdw0000gn/T/python-build.20201116195357.15726/Python-3.8.0/Lib/ensurepip/__init__.py", line 27, in _run_pip
    import pip._internal
  File "<frozen zipimport>", line 241, in load_module
  File "<frozen zipimport>", line 709, in _get_module_code
  File "<frozen zipimport>", line 570, in _get_data
zipimport.ZipImportError: can't decompress data; zlib not available
```

May it be a problem with the zlib? I installed that but it is not linked, if i try to link it (brew linlk zlib) the answer is: Will not link system shadow...

I tried to search for libz and give different paths to the command, bitz did not succeed...

```
find /usr -name "libz*"
/usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libzip.rb
/usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libzt.rb
/usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libzzip.rb
/usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libzdb.rb
/usr/local/var/homebrew/linked/libzip
/usr/local/lib/libzip.dylib
/usr/local/lib/pkgconfig/libzip.pc
/usr/local/lib/libzip.5.dylib
/usr/local/lib/cmake/libzip
/usr/local/lib/libzip.5.3.dylib
/usr/local/opt/libzip
/usr/local/Cellar/libzip
/usr/local/Cellar/libzip/1.7.3/.brew/libzip.rb
/usr/local/Cellar/libzip/1.7.3/lib/libzip.dylib
/usr/local/Cellar/libzip/1.7.3/lib/pkgconfig/libzip.pc
/usr/local/Cellar/libzip/1.7.3/lib/libzip.5.dylib
/usr/local/Cellar/libzip/1.7.3/lib/cmake/libzip
/usr/local/Cellar/libzip/1.7.3/lib/cmake/libzip/libzip-config-version.cmake
/usr/local/Cellar/libzip/1.7.3/lib/cmake/libzip/libzip-targets-release.cmake
/usr/local/Cellar/libzip/1.7.3/lib/cmake/libzip/libzip-config.cmake
/usr/local/Cellar/libzip/1.7.3/lib/cmake/libzip/libzip-targets.cmake
/usr/local/Cellar/libzip/1.7.3/lib/libzip.5.3.dylib
/usr/local/Cellar/libzip/1.7.3/share/man/man3/libzip.3
/usr/local/share/man/man3/libzip.3
```

I tried:

- -L/usr/local/lib
- -L/usr/local/Cellar/libzip/1.7.3/lib

but no luck. <grrrrrrr />

Any suggestions? All the best!

Reporting in with a similar issue trying to install `3.4.3`. I've tired several of the suggested fixes including setting the XCode SDK, trying several of the suggested bash scripts and setting CFLAGS and LFLAGS, to no avail. Running the install command provides the following error:

```
➜ pyenv install 3.4.3
python-build: use openssl from homebrew
python-build: use readline from homebrew
Downloading Python-3.4.3.tar.xz...
-> https://www.python.org/ftp/python/3.4.3/Python-3.4.3.tar.xz
Installing Python-3.4.3...
python-build: use readline from homebrew
python-build: use zlib from xcode sdk

BUILD FAILED (OS X 11.0.1 using python-build 20180424)

Inspect or clean up the working tree at /var/folders/fd/d66fbjt52158hypywt3hqrxh0000gn/T/python-build.20201120135412.16826
Results logged to /var/folders/fd/d66fbjt52158hypywt3hqrxh0000gn/T/python-build.20201120135412.16826.log

Last 10 log lines:
		-o Python/sysmodule.o ./Python/sysmodule.c
        int res = getentropy(buffer, len);
                  ^
Python/random.c:81:19: note: did you mean 'py_getentropy'?
Python/random.c:77:1: note: 'py_getentropy' declared here
py_getentropy(unsigned char *buffer, Py_ssize_t size, int fatal)
^
1 error generated.
make: *** [Python/random.o] Error 1
make: *** Waiting for unfinished jobs....
```

Unfortunately Python 3.4.3 is a hard dependency for me at work, so if I'm unable to install it I may have to downgrade to Catalina.

``
brew remove binutils
``

Fixed my issue where I was getting link error:

``
Last 10 log lines:
  "__PyRuntime", referenced from:
      _check_preinit_isolated_config in _testembed.o
  "__Py_Dealloc", referenced from:
      _test_open_code_hook in _testembed.o
      __audit_hook_run in _testembed.o
  "__Py_InitializeMain", referenced from:
      _test_init_main in _testembed.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [Programs/_testembed] Error 1
``

What's the solution for this?

I couldn't install Python 3.6.11 using Xcode 12.2 (12B45b) on macOS 11.0.1 (20B29), Intel.

```
Last 10 log lines:
        ret = sendfile(in, out, offset, &sbytes, &sf, flags);
              ^
./Modules/posixmodule.c:10432:5: warning: code will never be executed [-Wunreachable-code]
    Py_FatalError("abort() called from Python code didn't abort!");
    ^~~~~~~~~~~~~
clang -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include -O2  -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include -O2  -std=c99 -Wextra -Wno-unused-result -Wno-unused-parameter -Wno-missing-field-initializers -Wstrict-prototypes   -I. -I./Include -I/usr/local/opt/readline/include -I/usr/local/opt/readline/include -I/usr/local/opt/openssl@1.1/include -I/Users/i011842/.pyenv/versions/3.6.11/include  -I/usr/local/opt/zlib/include -I/usr/local/opt/sqlite/include -I/usr/local/opt/readline/include -I/usr/local/opt/readline/include -I/usr/local/opt/openssl@1.1/include -I/Users/i011842/.pyenv/versions/3.6.11/include  -I/usr/local/opt/zlib/include -I/usr/local/opt/sqlite/include  -DPy_BUILD_CORE  -c ./Modules/pwdmodule.c -o Modules/pwdmodule.o
1 warning generated.
1 warning and 1 error generated.

```

> @pardeep-singh Thank you, it works perfectly! You saved me a lot of trouble.
> @adib try the solution he posted. I had exactly the same problem and it worked for me. If you need a different version, just run
> 
> `CFLAGS="-I$(brew --prefix openssl)/include -I$(brew --prefix bzip2)/include -I$(brew --prefix readline)/include -I$(xcrun --show-sdk-path)/usr/include" LDFLAGS="-L$(brew --prefix openssl)/lib -L$(brew --prefix readline)/lib -L$(brew --prefix zlib)/lib -L$(brew --prefix bzip2)/lib" pyenv install --patch 3.6.12 < <(curl -sSL https://github.com/python/cpython/commit/8ea6353.patch\?full_index\=1)`

It did not work for me.

```log
...
python-build: use zlib from xcode sdk

BUILD FAILED (OS X 11.0.1 using python-build 20180424)

Inspect or clean up the working tree at /var/folders/v1/p2fcwmf52k589vx8g5st_3zw0000gn/T/python-build.20201202163059.1132
Results logged to /var/folders/v1/p2fcwmf52k589vx8g5st_3zw0000gn/T/python-build.20201202163059.1132.log

Last 10 log lines:
  File "/private/var/folders/v1/p2fcwmf52k589vx8g5st_3zw0000gn/T/python-build.20201202163059.1132/Python-3.6.12/Lib/ensurepip/__main__.py", line 5, in <module>
    sys.exit(ensurepip._main())
  File "/private/var/folders/v1/p2fcwmf52k589vx8g5st_3zw0000gn/T/python-build.20201202163059.1132/Python-3.6.12/Lib/ensurepip/__init__.py", line 204, in _main
    default_pip=args.default_pip,
  File "/private/var/folders/v1/p2fcwmf52k589vx8g5st_3zw0000gn/T/python-build.20201202163059.1132/Python-3.6.12/Lib/ensurepip/__init__.py", line 117, in _bootstrap
    return _run_pip(args + [p[0] for p in _PROJECTS], additional_paths)
  File "/private/var/folders/v1/p2fcwmf52k589vx8g5st_3zw0000gn/T/python-build.20201202163059.1132/Python-3.6.12/Lib/ensurepip/__init__.py", line 27, in _run_pip
    import pip._internal
zipimport.ZipImportError: can't decompress data; zlib not available
make: *** [install] Error 1
```

I have brew zlib installed. `brew --prefix zlib` returns `/usr/local/Cellar/zlib/1.2.11`. Why?

@llach The command by @pardeep-singh didn't work to install Python 3.6.11 via `pyenv`.

```
CFLAGS="-I$(brew --prefix openssl)/include -I$(brew --prefix bzip2)/include -I$(brew --prefix readline)/include -I$(xcrun --show-sdk-path)/usr/include" LDFLAGS="-L$(brew --prefix openssl)/lib -L$(brew --prefix readline)/lib -L$(brew --prefix zlib)/lib -L$(brew --prefix bzip2)/lib" pyenv install --patch 3.6.11 < <(curl -sSL https://github.com/python/cpython/commit/8ea6353.patch\?full_index\=1)

```

Result:

```
BUILD FAILED (OS X 10.16 using python-build 20180424)

Inspect or clean up the working tree at /var/folders/zf/5l7jzgw92jj9pdytfwfpbt_80000gp/T/python-build.20201202185939.36933
Results logged to /var/folders/zf/5l7jzgw92jj9pdytfwfpbt_80000gp/T/python-build.20201202185939.36933.log

Last 10 log lines:
  File "/private/var/folders/zf/5l7jzgw92jj9pdytfwfpbt_80000gp/T/python-build.20201202185939.36933/Python-3.6.11/Lib/ensurepip/__main__.py", line 5, in <module>
    sys.exit(ensurepip._main())
  File "/private/var/folders/zf/5l7jzgw92jj9pdytfwfpbt_80000gp/T/python-build.20201202185939.36933/Python-3.6.11/Lib/ensurepip/__init__.py", line 204, in _main
    default_pip=args.default_pip,
  File "/private/var/folders/zf/5l7jzgw92jj9pdytfwfpbt_80000gp/T/python-build.20201202185939.36933/Python-3.6.11/Lib/ensurepip/__init__.py", line 117, in _bootstrap
    return _run_pip(args + [p[0] for p in _PROJECTS], additional_paths)
  File "/private/var/folders/zf/5l7jzgw92jj9pdytfwfpbt_80000gp/T/python-build.20201202185939.36933/Python-3.6.11/Lib/ensurepip/__init__.py", line 27, in _run_pip
    import pip._internal
zipimport.ZipImportError: can't decompress data; zlib not available
make: *** [install] Error 1

```

I am using the official Big Sur release, macOS 11.1, and facing the same issue, but none of the above options are working.

Here are some relevant outputs:

```
 ~ clang --version
Apple clang version 12.0.0 (clang-1200.0.32.28)
Target: x86_64-apple-darwin20.2.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```


```
 ~ sw_vers
ProductName:	macOS
ProductVersion:	11.1
BuildVersion:	20C69
```

```
~ pyenv install 3.8.0
python-build: use openssl@1.1 from homebrew
python-build: use readline from homebrew
Installing Python-3.8.0...
python-build: use readline from homebrew
python-build: use zlib from xcode sdk

BUILD FAILED (OS X 11.1 using python-build 20180424)

Inspect or clean up the working tree at /var/folders/s8/x55yhgyx6t7c5rk_3gdqvfnm0000gn/T/python-build.20201224144959.34788
Results logged to /var/folders/s8/x55yhgyx6t7c5rk_3gdqvfnm0000gn/T/python-build.20201224144959.34788.log

Last 10 log lines:
                                                     ^
clang -c -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include   -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include   -std=c99 -Wextra -Wno-unused-result -Wno-unused-parameter -Wno-missing-field-initializers -Wstrict-prototypes -Werror=implicit-function-declaration  -I./Include/internal  -I. -I./Include -I/usr/local/opt/readline/include -I/usr/local/opt/readline/include -I/Users/justachetan/.pyenv/versions/3.8.0/include  -I/usr/local/opt/readline/include -I/usr/local/opt/readline/include -I/Users/justachetan/.pyenv/versions/3.8.0/include   -DPy_BUILD_CORE -o Modules/gcmodule.o Modules/gcmodule.c
clang -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include   -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include   -std=c99 -Wextra -Wno-unused-result -Wno-unused-parameter -Wno-missing-field-initializers -Wstrict-prototypes -Werror=implicit-function-declaration  -I./Include/internal  -I. -I./Include -I/usr/local/opt/readline/include -I/usr/local/opt/readline/include -I/Users/justachetan/.pyenv/versions/3.8.0/include  -I/usr/local/opt/readline/include -I/usr/local/opt/readline/include -I/Users/justachetan/.pyenv/versions/3.8.0/include   -DPy_BUILD_CORE_BUILTIN  -DPy_BUILD_CORE_BUILTIN -I./Include/internal -c ./Modules/posixmodule.c -o Modules/posixmodule.o
./Modules/posixmodule.c:9084:15: error: implicit declaration of function 'sendfile' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
        ret = sendfile(in, out, offset, &sbytes, &sf, flags);
              ^
1 error generated.
make: *** [Modules/posixmodule.o] Error 1
make: *** Waiting for unfinished jobs....
```

[python-build-380.log](https://github.com/pyenv/pyenv/files/5739023/python-build.log)

Getting similar results for Python 3.6.6 as well: 

```
~ pyenv install 3.6.6
python-build: use openssl from homebrew
python-build: use readline from homebrew
Installing Python-3.6.6...
python-build: use readline from homebrew
python-build: use zlib from xcode sdk

BUILD FAILED (OS X 11.1 using python-build 20180424)

Inspect or clean up the working tree at /var/folders/s8/x55yhgyx6t7c5rk_3gdqvfnm0000gn/T/python-build.20201224145652.44797
Results logged to /var/folders/s8/x55yhgyx6t7c5rk_3gdqvfnm0000gn/T/python-build.20201224145652.44797.log

Last 10 log lines:
./Modules/posixmodule.c:8210:15: error: implicit declaration of function 'sendfile' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
        ret = sendfile(in, out, offset, &sbytes, &sf, flags);
              ^
./Modules/posixmodule.c:10432:5: warning: code will never be executed [-Wunreachable-code]
    Py_FatalError("abort() called from Python code didn't abort!");
    ^~~~~~~~~~~~~
clang -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include   -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include   -std=c99 -Wextra -Wno-unused-result -Wno-unused-parameter -Wno-missing-field-initializers -Wstrict-prototypes   -I. -I./Include -I/usr/local/opt/readline/include -I/usr/local/opt/readline/include -I/usr/local/opt/openssl@1.1/include -I/Users/justachetan/.pyenv/versions/3.6.6/include  -I/usr/local/opt/readline/include -I/usr/local/opt/readline/include -I/usr/local/opt/openssl@1.1/include -I/Users/justachetan/.pyenv/versions/3.6.6/include   -DPy_BUILD_CORE  -c ./Modules/pwdmodule.c -o Modules/pwdmodule.o
1 warning and 1 error generated.
make: *** [Modules/posixmodule.o] Error 1
make: *** Waiting for unfinished jobs....
```

[python-build-366.log](https://github.com/pyenv/pyenv/files/5739045/python-build-366.log)

I have tried the approaches given by @franz-josef-kaiser @pardeep-singh @ahmadmoawad still not working.

Any help on this would be really appreciated.

> I am using the official Big Sur release, macOS 11.1, and facing the same issue, but none of the above options are working.
> 
> Here are some relevant outputs:
> 
> ```
>  ~ clang --version
> Apple clang version 12.0.0 (clang-1200.0.32.28)
> Target: x86_64-apple-darwin20.2.0
> Thread model: posix
> InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
> ```
> 
> ```
>  ~ sw_vers
> ProductName:	macOS
> ProductVersion:	11.1
> BuildVersion:	20C69
> ```
> 
> ```
> ~ pyenv install 3.8.0
> python-build: use openssl@1.1 from homebrew
> python-build: use readline from homebrew
> Installing Python-3.8.0...
> python-build: use readline from homebrew
> python-build: use zlib from xcode sdk
> 
> BUILD FAILED (OS X 11.1 using python-build 20180424)
> 
> Inspect or clean up the working tree at /var/folders/s8/x55yhgyx6t7c5rk_3gdqvfnm0000gn/T/python-build.20201224144959.34788
> Results logged to /var/folders/s8/x55yhgyx6t7c5rk_3gdqvfnm0000gn/T/python-build.20201224144959.34788.log
> 
> Last 10 log lines:
>                                                      ^
> clang -c -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include   -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include   -std=c99 -Wextra -Wno-unused-result -Wno-unused-parameter -Wno-missing-field-initializers -Wstrict-prototypes -Werror=implicit-function-declaration  -I./Include/internal  -I. -I./Include -I/usr/local/opt/readline/include -I/usr/local/opt/readline/include -I/Users/justachetan/.pyenv/versions/3.8.0/include  -I/usr/local/opt/readline/include -I/usr/local/opt/readline/include -I/Users/justachetan/.pyenv/versions/3.8.0/include   -DPy_BUILD_CORE -o Modules/gcmodule.o Modules/gcmodule.c
> clang -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include   -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include   -std=c99 -Wextra -Wno-unused-result -Wno-unused-parameter -Wno-missing-field-initializers -Wstrict-prototypes -Werror=implicit-function-declaration  -I./Include/internal  -I. -I./Include -I/usr/local/opt/readline/include -I/usr/local/opt/readline/include -I/Users/justachetan/.pyenv/versions/3.8.0/include  -I/usr/local/opt/readline/include -I/usr/local/opt/readline/include -I/Users/justachetan/.pyenv/versions/3.8.0/include   -DPy_BUILD_CORE_BUILTIN  -DPy_BUILD_CORE_BUILTIN -I./Include/internal -c ./Modules/posixmodule.c -o Modules/posixmodule.o
> ./Modules/posixmodule.c:9084:15: error: implicit declaration of function 'sendfile' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
>         ret = sendfile(in, out, offset, &sbytes, &sf, flags);
>               ^
> 1 error generated.
> make: *** [Modules/posixmodule.o] Error 1
> make: *** Waiting for unfinished jobs....
> ```
> 
> [python-build-380.log](https://github.com/pyenv/pyenv/files/5739023/python-build.log)
> 
> Getting similar results for Python 3.6.6 as well:
> 
> ```
> ~ pyenv install 3.6.6
> python-build: use openssl from homebrew
> python-build: use readline from homebrew
> Installing Python-3.6.6...
> python-build: use readline from homebrew
> python-build: use zlib from xcode sdk
> 
> BUILD FAILED (OS X 11.1 using python-build 20180424)
> 
> Inspect or clean up the working tree at /var/folders/s8/x55yhgyx6t7c5rk_3gdqvfnm0000gn/T/python-build.20201224145652.44797
> Results logged to /var/folders/s8/x55yhgyx6t7c5rk_3gdqvfnm0000gn/T/python-build.20201224145652.44797.log
> 
> Last 10 log lines:
> ./Modules/posixmodule.c:8210:15: error: implicit declaration of function 'sendfile' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
>         ret = sendfile(in, out, offset, &sbytes, &sf, flags);
>               ^
> ./Modules/posixmodule.c:10432:5: warning: code will never be executed [-Wunreachable-code]
>     Py_FatalError("abort() called from Python code didn't abort!");
>     ^~~~~~~~~~~~~
> clang -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include   -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include   -std=c99 -Wextra -Wno-unused-result -Wno-unused-parameter -Wno-missing-field-initializers -Wstrict-prototypes   -I. -I./Include -I/usr/local/opt/readline/include -I/usr/local/opt/readline/include -I/usr/local/opt/openssl@1.1/include -I/Users/justachetan/.pyenv/versions/3.6.6/include  -I/usr/local/opt/readline/include -I/usr/local/opt/readline/include -I/usr/local/opt/openssl@1.1/include -I/Users/justachetan/.pyenv/versions/3.6.6/include   -DPy_BUILD_CORE  -c ./Modules/pwdmodule.c -o Modules/pwdmodule.o
> 1 warning and 1 error generated.
> make: *** [Modules/posixmodule.o] Error 1
> make: *** Waiting for unfinished jobs....
> ```
> 
> [python-build-366.log](https://github.com/pyenv/pyenv/files/5739045/python-build-366.log)
> 
> I have tried the approaches given by @franz-josef-kaiser @pardeep-singh @ahmadmoawad still not working.
> 
> Any help on this would be really appreciated.

Please let me know if you get it to work. I've tried everything listed above as well

newlines, yeah great :) :
ubuntus-MBP:ProperTree ubuntu$ LDFLAGS="\                      
> -L$(brew --prefix zlib)/lib \
> -L$(brew --prefix bzip2)/lib \
> " pyenv install 3.9.0
python-build: use openssl@1.1 from homebrew
python-build: use readline from homebrew
Downloading Python-3.9.0.tar.xz...
-> https://www.python.org/ftp/python/3.9.0/Python-3.9.0.tar.xz
Installing Python-3.9.0...
python-build: use readline from homebrew
python-build: use zlib from xcode sdk

BUILD FAILED (OS X 11.1 using python-build 20180424)

Inspect or clean up the working tree at /var/folders/nb/d_v4kfms651cqg8tv7sdnwm40000gn/T/python-build.20210104024021.15778
Results logged to /var/folders/nb/d_v4kfms651cqg8tv7sdnwm40000gn/T/python-build.20210104024021.15778.log

Last 10 log lines:
checking for --enable-universalsdk... no
checking for --with-universal-archs... no
checking MACHDEP... "darwin"
checking for gcc... clang
checking whether the C compiler works... no
configure: error: in `/var/folders/nb/d_v4kfms651cqg8tv7sdnwm40000gn/T/python-build.20210104024021.15778/Python-3.9.0':
configure: error: C compiler cannot create executables
See `config.log' for more details
configure: WARNING: cache variable ac_cv_env_LDFLAGS_value contains a newline
make: *** No targets specified and no makefile found.  Stop.

I saw this when installing bzip with brew:
_If you need to have bzip2 first in your PATH run:
  echo 'export PATH="/usr/local/opt/bzip2/bin:$PATH"' >> /Users/ubuntu/.bash_profile

For compilers to find bzip2 you may need to set:
  export LDFLAGS="-L/usr/local/opt/bzip2/lib"
  export CPPFLAGS="-I/usr/local/opt/bzip2/include"_

I did not install zlib with brew and didn't want to since it's in the OSX Developer tools.
Based on that I came up with:

`brew install bzip2
`Xcode developers tools installed 
add to ~/.bash_profile :
```
export LDFLAGS="-L $(xcrun --show-sdk-path)/usr/lib -L brew --prefix bzip2/lib"
export CFLAGS="-L $(xcrun --show-sdk-path)/usr/include -L brew --prefix bzip2/include"
```

then just : 

```
ubuntus-MBP:~ ubuntu$ pyenv install 3.9.0
python-build: use openssl@1.1 from homebrew
python-build: use readline from homebrew
Downloading Python-3.9.0.tar.xz...
-> https://www.python.org/ftp/python/3.9.0/Python-3.9.0.tar.xz
Installing Python-3.9.0...
python-build: use readline from homebrew
python-build: use zlib from xcode sdk
Installed Python-3.9.0 to /Users/ubuntu/.pyenv/versions/3.9.0
```

I prefer this because it should work for other versions and similar commands and I don't have to remember or put more stuff on the command line than usual.

I hope pyenv will be modded to detect the library locations in Big Sur.
Thanks for the examples and hints to everyone!

Referencing the above, I've been trying unsuccessfully for a native compile using the following:

> (rosetta works without any issue with this config - goal is both)

## .zshrc
```shell

ARCH=`arch`
if [[ "${ARCH}"  == "arm64" ]]; then
    export PYENV_ROOT="/Users/rob/.pyenv/arm64"
else
    export PYENV_ROOT="/Users/rob/.pyenv/rosetta"
fi
PYENV_BIN="$PYENV_ROOT/bin"
export PYENV_SHELL=zsh
export PYTHONPATH=$PYENV_ROOT/shims


SDK_PATH="$(xcrun --show-sdk-path)"
export CPATH="${SDK_PATH}/usr/include"
export CFLAGS="-I${SDK_PATH}/usr/include $CFLAGS"
export CFLAGS="-I${SDK_PATH}/usr/include/sasl $CFLAGS"
export LDFLAGS="-L${SDK_PATH}/usr/lib $LDFLAGS"


if [[ "${ARCH}"  == "arm64" ]]; then
    PREFIX="/opt/homebrew/opt"
else
    PREFIX="/usr/local/opt"
fi


ZLIB="${PREFIX}/zlib"
BZIP2="${PREFIX}/bzip2"
READLINE="${PREFIX}/readline"
SQLITE="${PREFIX}/sqlite"    
OPENSSL="${PREFIX}/openssl@1.1"
PGSQL="${PREFIX}/postgresql@10"

LIBS=('ZLIB' 'BZIP2' 'READLINE' 'SQLITE' 'OPENSSL' 'PGSQL')

for LIB in $LIBS; do

	BINDIR="${(P)LIB}/bin"
	if [ -d "${BINDIR}" ]; then
		export PATH="${BINDIR}:$PATH"
	fi

	LIBDIR="${(P)LIB}/lib"
	if [ -d "${LIBDIR}" ]; then
		export LDFLAGS="-L${LIBDIR} $LDFLAGS"
		export DYLD_LIBRARY_PATH="${LIBDIR}:$DYLD_LIBRARY_PATH"
		PKGCFGDIR="${LIBDIR}/pkgconfig"
		if [ -d "${PKGCFGDIR}" ]; then
			export PKG_CONFIG_PATH="${PKGCFGDIR} $PKG_CONFIG_PATH"
		fi
	fi

	INCDIR="${(P)LIB}/include"
	if [ -d "${INCDIR}" ]; then
		export CFLAGS="-I${INCDIR} $CFLAGS" 
	fi

done

export CPPFLAGS="$CFLAGS"
export CXXFLAGS=$CFLAGS


eval "$(pyenv init -)"

```

## Result:
```log
configure: error: Unexpected output of 'arch' on OSX
```

```
brew install zlib
brew install sqlite
brew install bzip2
brew install libiconv
brew install libzip
`LDFLAGS="-L$(brew --prefix zlib)/lib -L$(brew --prefix bzip2)/lib" pyenv install --patch 3.4.10 < <(curl -sSL https://github.com/python/cpython/commit/8ea6353.patch\?full_index\=1)`
```
For me doesn't work, and all other versions too... The result is:

```% pyenv install 3.7.2
python-build: use openssl@1.1 from homebrew
python-build: use readline from homebrew
Downloading Python-3.7.2.tar.xz...
-> https://www.python.org/ftp/python/3.7.2/Python-3.7.2.tar.xz
Installing Python-3.7.2...
python-build: use readline from homebrew
python-build: use zlib from xcode sdk

BUILD FAILED (OS X 11.1 using python-build 20180424)

Inspect or clean up the working tree at /var/folders/0b/w9y7p9zx1xx89l_q1t1sk0gm0000gp/T/python-build.20210115172508.52082
Results logged to /var/folders/0b/w9y7p9zx1xx89l_q1t1sk0gm0000gp/T/python-build.20210115172508.52082.log

Last 10 log lines:
clang -c -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include   -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include   -std=c99 -Wextra -Wno-unused-result -Wno-unused-parameter -Wno-missing-field-initializers -Wstrict-prototypes -Werror=implicit-function-declaration   -I. -I./Include -I/usr/local/opt/readline/include -I/usr/local/opt/readline/include -I/Users/andrey/.pyenv/versions/3.7.2/include  -I/usr/local/opt/readline/include -I/usr/local/opt/readline/include -I/Users/andrey/.pyenv/versions/3.7.2/include   -DPy_BUILD_CORE -o Modules/main.o Modules/main.c
clang -c -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include   -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include   -std=c99 -Wextra -Wno-unused-result -Wno-unused-parameter -Wno-missing-field-initializers -Wstrict-prototypes -Werror=implicit-function-declaration   -I. -I./Include -I/usr/local/opt/readline/include -I/usr/local/opt/readline/include -I/Users/andrey/.pyenv/versions/3.7.2/include  -I/usr/local/opt/readline/include -I/usr/local/opt/readline/include -I/Users/andrey/.pyenv/versions/3.7.2/include   -DPy_BUILD_CORE -o Modules/gcmodule.o Modules/gcmodule.c
1 warning generated.
clang -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include   -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include   -std=c99 -Wextra -Wno-unused-result -Wno-unused-parameter -Wno-missing-field-initializers -Wstrict-prototypes -Werror=implicit-function-declaration   -I. -I./Include -I/usr/local/opt/readline/include -I/usr/local/opt/readline/include -I/Users/andrey/.pyenv/versions/3.7.2/include  -I/usr/local/opt/readline/include -I/usr/local/opt/readline/include -I/Users/andrey/.pyenv/versions/3.7.2/include   -DPy_BUILD_CORE_BUILTIN  -DPy_BUILD_CORE -c ./Modules/posixmodule.c -o Modules/posixmodule.o
./Modules/posixmodule.c:8409:15: error: implicit declaration of function 'sendfile' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
        ret = sendfile(in, out, offset, &sbytes, &sf, flags);
              ^
1 error generated.
make: *** [Modules/posixmodule.o] Error 1
make: *** Waiting for unfinished jobs....
andrey@Andrey-MBP2017 ~ % LDFLAGS="-L$(brew --prefix zlib)/lib -L$(brew --prefix bzip2)/lib" pyenv install --patch 3.4.10 < <(curl -sSL https://github.com/python/cpython/commit/8ea6353.patch\?full_index\=1)
```
Any ideas?

I've tried all of the suggested solutions and have yet to get any of them working. When I try `LDFLAGS="-L$(brew --prefix zlib)/lib -L$(brew --prefix bzip2)/lib" pyenv install 3.8.6` for example as @VigneshChennai suggested, I get the following output:

```
python-build: use openssl@1.1 from homebrew
python-build: use readline from homebrew
Downloading Python-3.8.6.tar.xz...
-> https://www.python.org/ftp/python/3.8.6/Python-3.8.6.tar.xz
Installing Python-3.8.6...
python-build: use readline from homebrew
python-build: use zlib from xcode sdk

BUILD FAILED (OS X 11.1 using python-build 20180424)

Inspect or clean up the working tree at /var/folders/j4/0sbp2d8n71n7m2snn2_0ns1c0000gn/T/python-build.20210122001428.80486
Results logged to /var/folders/j4/0sbp2d8n71n7m2snn2_0ns1c0000gn/T/python-build.20210122001428.80486.log

Last 10 log lines:
                                    ^
./Include/cpython/tupleobject.h:23:28: note: expanded from macro '_PyTuple_CAST'
#define _PyTuple_CAST(op) (assert(PyTuple_Check(op)), (PyTupleObject *)(op))
                           ^
Objects/abstract.c:2129:9: error: implicit declaration of function 'assert' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
        assert(len >= 0 || PyErr_Occurred());
        ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
2 warnings and 20 errors generated.
make: *** [Objects/abstract.o] Error 1
```

I am not seeing anyone else running into the `assert` error specifically, so is there some other issue that I'm missing?

I have tried three solutions here and none of them can install python 3.5.2

https://github.com/pyenv/pyenv/issues/1643#issuecomment-687643468
```
Python/random.c:97:19: note: did you mean 'py_getentropy'?
Python/random.c:89:1: note: 'py_getentropy' declared here
py_getentropy(unsigned char *buffer, Py_ssize_t size, int fatal)
^
Python/random.c:106:19: error: implicit declaration of function 'getentropy' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
            res = getentropy(buffer, len);
```
https://github.com/pyenv/pyenv/issues/1643#issuecomment-753015504
```
make: *** [Modules/posixmodule.o] Error 1
make: *** Waiting for unfinished jobs....
clang: warning: argument unused during compilation: '-L/usr/local/Cellar/zlib/1.2.11/include' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-L/usr/local/opt/bzip2/include' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-L/usr/local/Cellar/zlib/1.2.11/include' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-L/usr/local/opt/bzip2/include' [-Wunused-command-line-argument]
```
and 
https://github.com/pyenv/pyenv/issues/1643#issuecomment-753429938
```
checksum mismatch: Python-3.5.2.tar.xz (file is corrupt)
expected 0010f56100b9b74259ebcd5d4b295a32324b58b517403a10d1a2aa7cb22bca40, got ff20c15722cbc98ebb9198e98f2eaea577670410c4b1543fd095716071709455
```

Operating System: `macOS Big Sur version 11.1 (20C69)` 

Like @htp I also was able to install Pythons `3.8.5` by omitting the patch. But then I tried the same for `3.5.10` and build failed with this throw 
```
BUILD FAILED (OS X 11.1 using python-build 20180424)

Inspect or clean up the working tree at /var/folders/lr/r58bs_3n2014mps32rflkmxc0000gn/T/python-build.20210123115655.87782
Results logged to /var/folders/lr/r58bs_3n2014mps32rflkmxc0000gn/T/python-build.20210123115655.87782.log

Last 10 log lines:
./Modules/posixmodule.c:10614:5: warning: code will never be executed [-Wunreachable-code]
    Py_FatalError("abort() called from Python code didn't abort!");
    ^~~~~~~~~~~~~
./Modules/posixmodule.c:11345:9: warning: code will never be executed [-Wunreachable-code]
        posix_error();
        ^~~~~~~~~~~
11 warnings and 1 error generated.
make: *** [Modules/posixmodule.o] Error 1
make: *** Waiting for unfinished jobs....
1 warning generated.
```

So I finally managed to install `3.5.10` by using the patch thanks to @PritamDutt  comment 
```bash
CFLAGS="-I$(brew --prefix openssl)/include -I$(brew --prefix readline)/include -I$(xcrun --show-sdk-path)/usr/include" \
LDFLAGS="-L$(brew --prefix openssl)/lib -L$(brew --prefix readline)/lib -L$(xcrun --show-sdk-path)/usr/lib" \
pyenv install --patch 3.5.10 < <(curl -sSL https://github.com/python/cpython/commit/8ea6353.patch)
```

If anyone is trying to install 2.7.12 on OSX Catalina+ and sees `Python/random.c:123:19: error: implicit declaration of function 'getentropy' is invalid in C99 [-Werror,-Wimplicit-function-declaration]`

This has been patched in [2.7.13, 3.5.3, 3.6.0b4, and 3.7.0](https://bugs.python.org/issue28676)

> I tried building using those inputs to my .zshrc and still got an error:
> 
> ```
> andytriboletti@andys-mac ~ % CFLAGS="-I$(brew --prefix openssl)/include -I$(brew --prefix bzip2)/include -I$(brew --prefix readline)/include -I$(xcrun --show-sdk-path)/usr/include" LDFLAGS="-L$(brew --prefix openssl)/lib -L$(brew --prefix readline)/lib -L$(brew --prefix zlib)/lib -L$(brew --prefix bzip2)/lib" \
> pyenv install --patch 3.8.3 < <(curl -sSL https://github.com/python/cpython/commit/8ea6353.patch\?full_index\=1)
> 
> python-build: use openssl@1.1 from homebrew
> python-build: use readline from homebrew
> Downloading Python-3.8.3.tar.xz...
> -> https://www.python.org/ftp/python/3.8.3/Python-3.8.3.tar.xz
> Installing Python-3.8.3...
> patching file Misc/NEWS.d/next/macOS/2020-06-24-13-51-57.bpo-41100.mcHdc5.rst
> patching file configure
> Hunk #1 succeeded at 3398 (offset -28 lines).
> patching file configure.ac
> Hunk #1 succeeded at 498 (offset -12 lines).
> python-build: use readline from homebrew
> python-build: use zlib from xcode sdk
> 
> BUILD FAILED (OS X 11.0 using python-build 20180424)
> 
> Inspect or clean up the working tree at /var/folders/1r/bdbwv9r53ls_80hpk1t6fvlm0000gn/T/python-build.20200715094722.77820
> Results logged to /var/folders/1r/bdbwv9r53ls_80hpk1t6fvlm0000gn/T/python-build.20200715094722.77820.log
> 
> Last 10 log lines:
> checking size of _Bool... 1
> checking size of off_t... 8
> checking whether to enable large file support... no
> checking size of time_t... 8
> checking for pthread_t... yes
> checking size of pthread_t... 8
> checking size of pthread_key_t... 8
> checking whether pthread_key_t is compatible with int... no
> configure: error: Unexpected output of 'arch' on OSX
> make: *** No targets specified and no makefile found.  Stop.
> andytriboletti@andys-mac ~ % arch
> arm64
> ```

I'm just writing down my experiences here, maybe it is going to be helpful to somebody else too. So I'm running on Big Sur on a Mac **M1 chip** and I had the exact same error output when I tried to install python 3.6.12. The key for me to add `arch -x86_64` just before the `pyenv install` part of the command.

```
CFLAGS="-I$(brew --prefix openssl)/include -I$(brew --prefix bzip2)/include -I$(brew --prefix readline)/include -I$(xcrun --show-sdk-path)/usr/include" LDFLAGS="-L$(brew --prefix openssl)/lib -L$(brew --prefix readline)/lib -L$(brew --prefix zlib)/lib -L$(brew --prefix bzip2)/lib" arch -x86_64 pyenv install --patch 3.6.12 < <(curl -sSL https://github.com/python/cpython/commit/8ea6353.patch\?full_index\=1)
```

Well...I'm having this problem too...I'm using an intel mac.
```zsh
~/test 14s
❯ pyenv install 3.8.8
python-build: use openssl@1.1 from homebrew
python-build: use readline from homebrew
Downloading Python-3.8.8.tar.xz...
-> https://www.python.org/ftp/python/3.8.8/Python-3.8.8.tar.xz
error: failed to download Python-3.8.8.tar.xz

BUILD FAILED (OS X 11.2.3 using python-build 20180424)
```

> Well...I'm having this problem too...I'm using an intel mac.
> 
> ```shell
> ~/test 14s
> ❯ pyenv install 3.8.8
> python-build: use openssl@1.1 from homebrew
> python-build: use readline from homebrew
> Downloading Python-3.8.8.tar.xz...
> -> https://www.python.org/ftp/python/3.8.8/Python-3.8.8.tar.xz
> error: failed to download Python-3.8.8.tar.xz
> 
> BUILD FAILED (OS X 11.2.3 using python-build 20180424)
> ```
> 
> I can download the file by just typing the link to the browser. `wget` and `curl` should be properly installed by `brew`.

Comment on my own comment. It actually works now. It seems like there was some downloading issue happening and I couldn't download the Python-3.8.8.tar.xz.

### **ISSUE:**
I can't get ```pyenv install 3.8.2``` to work without ending in error.

**System Information:**
- OSX: 11.3.1(Big Sur)
- pyenv: 1.2.27
- zsh: 5.8

**Error message received:**
BUILD FAILED (OS X 11.3.1 using python-build 1.2.27-41-gddf952ab)
./Modules/posixmodule.c:9197:15: error: implicit declaration of function 'sendfile' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
        ret = sendfile(in, out, offset, &sbytes, &sf, flags);
              ^
1 error generated.

make: *** [Modules/posixmodule.o] Error 1


### **SOLUTION!!!!:**
Install with the --patch flag
```pyenv install --patch 3.8.2 < <(curl -sSL https://github.com/python/cpython/commit/8ea6353.patch)```

I would assume this is similar for other python version installs. Just swap the python install version and leave the patch URL the same.

--- a/plugins/python-build/share/python-build/patches/3.7.12/Python-3.7.12/0004-bpo-45405-Prevent-internal-configure-error-when-runn.patch
+++ b/plugins/python-build/share/python-build/patches/3.7.12/Python-3.7.12/0004-bpo-45405-Prevent-internal-configure-error-when-runn.patch
@@ -1,0 +1,43 @@
+From 720bb456dc711b0776bae837d1f9a0b10c28ddf2 Mon Sep 17 00:00:00 2001
+From: Ned Deily <nad@python.org>
+Date: Tue, 15 Mar 2022 03:18:39 -0400
+Subject: [PATCH] bpo-45405: Prevent internal configure error when running
+ configure with recent versions of clang. (GH-28845) (GH-31890)
+
+Change the configure logic to function properly on macOS when the compiler
+outputs a platform triplet for option --print-multiarch.
+The Apple Clang included with Xcode 13.3 now supports --print-multiarch
+causing configure to fail without this change.
+---
+ configure | 8 +++++---
+ 1 file changed, 5 insertions(+), 3 deletions(-)
+
+diff --git configure configure
+index 829dd69bb8..455481bc50 100755
+--- a/configure
++++ b/configure
+@@ -5183,9 +5183,6 @@ $as_echo "$as_me:
+ fi
+
+
+-MULTIARCH=$($CC --print-multiarch 2>/dev/null)
+-
+-
+ { $as_echo "$as_me:${as_lineno-$LINENO}: checking for the platform triplet based on compiler characteristics" >&5
+ $as_echo_n "checking for the platform triplet based on compiler characteristics... " >&6; }
+ cat >> conftest.c <<EOF
+@@ -5342,6 +5339,11 @@ $as_echo "none" >&6; }
+ fi
+ rm -f conftest.c conftest.out
+
++if test x$PLATFORM_TRIPLET != xdarwin; then
++  MULTIARCH=$($CC --print-multiarch 2>/dev/null)
++fi
++
++
+ if test x$PLATFORM_TRIPLET != x && test x$MULTIARCH != x; then
+   if test x$PLATFORM_TRIPLET != x$MULTIARCH; then
+     as_fn_error $? "internal configure error for the platform triplet, please file a bug report" "$LINENO" 5
+--
+2.35.1
+

